### ISSUE TITLE
Exception at GradientBuilderBase when trying to export pytorch vision transformer gradient graph for training 

### ISSUE THREAD

**Issue by pan-mic** [microsoft/onnxruntime#19038]
### Describe the issue

I am trying to extract the gradient graph of an image classifier model, so that i can use it with onnx runtime web to get the gradient over the image for a given class vector (for a project that requires a custom optimization loop). I'm using the pretrained models in torchvision and i've been successful with simpler models like Alexnet for which I do get the correct gradients when i load its onnx gradient graph in the web runtime. 
However when i apply the same process to vit_b_32 (and other vision transformer based models) i get an exception halfway through the conversion process at the python function call onnxblock._training_graph_utils.build_gradient_graph which leads to the c++ header file "orttraining\core\graph\gradient_builder_base.h:122 onnxruntime::training::GradientBuilderBase::O i < node_->OutputDefs().size() was false. "

### To reproduce

To simplify things i created a wrapper torch module with just the pretrained image classifier with the loss function as the final output in forward() and the image tensor on which i want to get the gradient as a parameter of the model named 'canvas'
Initially i used artifacts.generate_artifacts(...) but calling onnxblock._training_graph_utils.build_gradient_graph(...) seems to be enough to just get the gradient graph (at least it works fine for alexnet). For Vit_B_32 i get the same exception regardless of whether i setup the whole artifacts.generate_artifacts... pipeline or just call build_gradient_graph direclty.

 
```
import torch
import torch.nn as nn
from torchvision import models

import os

import onnx
import onnxruntime.training.onnxblock as onnxblock
from onnxruntime.training import artifacts


DEVICE = torch.device("cpu")

MODEL_NAME = 'vit_b_32'
IMGNET_MODEL = models.vit_b_32(weights = models.ViT_B_32_Weights.DEFAULT)
#MODEL_NAME = 'alexnet'
#IMGNET_MODEL = models.alexnet(weights = models.AlexNet_Weights.DEFAULT)
IMGNET_MODEL.to(DEVICE)

OUTPUT_FOLDER = 'D:/ML/SignLanguage/TEMP'
OUTPUT_MODEL = os.path.join(OUTPUT_FOLDER, f'{MODEL_NAME}_loss.onnx')
OUTPUT_GRAD_GRAPH = os.path.join(OUTPUT_FOLDER, f'{MODEL_NAME}_loss_grad.onnx')


EXPORT_OPTIONS = dict(input_names=['input'], output_names=['output'],
                    export_params=True,                             #train
                    verbose=False, opset_version=17,
                    do_constant_folding=False,                      #train
                    training = torch.onnx.TrainingMode.TRAINING,    #train
                    )

class ModelLossWrapper(nn.Module):
    def __init__(self):
        super().__init__()
        self.canvas = nn.Parameter(torch.randn([1, 3, 224, 224], dtype=torch.float32), requires_grad=True)
        self.visual = IMGNET_MODEL
        self.loss = nn.CosineSimilarity(dim=1)            

        for param in self.visual.parameters():
            param.requires_grad = False

    def forward(self, output_vec):
        vec = self.visual(self.canvas)
        loss =  -self.loss(vec, output_vec)
        return loss

modelWrapper = ModelLossWrapper()
vec_inp  = torch.randn(1, 1000, dtype=torch.float32).cpu()
torch.onnx.export(modelWrapper, vec_inp, OUTPUT_MODEL, EXPORT_OPTIONS)

# Load the forward only onnx model
model = onnx.load(OUTPUT_MODEL)

#print intputs and outputs
print("inputs: ")
for inp in model.graph.input:
    print(inp.name)
print("outputs: ")
outname = model.graph.output[0].name
for out in model.graph.output:
    print(out.name)

grad_params = set(['canvas'])
frozen_params = set([])
outputs = [outname]

grad_graph, eval_model = onnxblock._training_graph_utils.build_gradient_graph(model, grad_params, frozen_params, outputs)

#save gradient graph
onnx.save(grad_graph, OUTPUT_GRAD_GRAPH)``
```

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.13.1

### PyTorch Version

2.2.0.dev20231101+cu121

### Execution Provider

Default CPU

**Comment by prathikr** [comment]
@pan-mic can you retry with onnxruntime==1.16.3 and report whether you still see this issue?

**Comment by prathikr** [comment]
related issue: https://github.com/microsoft/onnxruntime-training-examples/issues/176

**Comment by pan-mic** [comment]
> @pan-mic can you retry with onnxruntime==1.16.3 and report whether you still see this issue?

@prathikr I installed onnxruntime-training-cpu 1.16.3 using the pip options from https://onnxruntime.ai/getting-started and still getting the same exception

**Comment by prathikr** [comment]
@pan-mic the root cause is usage of `opset_version=17` for exporting the model. `opset_version==17` introduced the [LayerNormalization operator](https://github.com/onnx/onnx/blob/main/docs/Operators.md#layernormalization) into the ONNX ecosystem but ORTModule defaults to `opset_version=15` which does not have this operator. Switching the opset_version to 15 and unravelling the EXPORT_OPTIONS dict during the export command (using **) results in successful exporting of both vit_b_32_loss.onnx and vit_b_32_loss_grad.onnx graphs using `opset_version==15`.

```python
EXPORT_OPTIONS = dict(input_names=['input'], output_names=['output'],
                    export_params=True,                             #train
                    verbose=False, opset_version=15,
                    do_constant_folding=False,                      #train
                    training = torch.onnx.TrainingMode.TRAINING,    #train
                    )

.
.
.

torch.onnx.export(modelWrapper, vec_inp, OUTPUT_MODEL, **EXPORT_OPTIONS)
```

We are currently in the process of upgrading ORTModule's opset from 15 to 20. You can track the progress from PR https://github.com/microsoft/onnxruntime/pull/19152
